storage/inkless/src/main/java/io/aiven/inkless/consume/FetchPlanner.java

```diff
@@ -70,51 +88,167 @@ public FetchPlanner(
+     *
+     * @param batchCoordinates map of partition to batch information
+     * @return list of planned fetch requests
+     */
+    List<ObjectFetchRequest> planJobs(final Map<TopicIdPartition, FindBatchResponse> batchCoordinates) {
+        // Group batches by object key and merge metadata
+        final Set<Map.Entry<String, List<MergedBatchRequest>>> objectKeysToRanges =
+            batchCoordinates.values().stream()
+                // Filter out responses with errors
+                .filter(findBatch -> findBatch.errors() == Errors.NONE)
+                .map(FindBatchResponse::batches)
+                .peek(batches -> metrics.recordFetchBatchSize(batches.size()))
+                .flatMap(List::stream)
+                .collect(Collectors.groupingBy(
+                    BatchInfo::objectKey,
+                    Collectors.mapping(MergedBatchRequest::create, Collectors.toList())
+                ))
+                .entrySet();
+
         metrics.recordFetchObjectsSize(objectKeysToRanges.size());
+
+        // Create fetch requests with aligned byte ranges
         return objectKeysToRanges.stream()
-            .flatMap(e ->
-                keyAlignment.align(e.getValue())
-                    .stream()
-                    .map(byteRange -> getCacheFetchJob(e.getKey(), byteRange)))
+            .flatMap(e -> createFetchRequests(e.getKey(), e.getValue()))
             .collect(Collectors.toList());
     }
 
-    private CacheFetchJob getCacheFetchJob(final String objectKey, final ByteRange byteRange) {
-        return new CacheFetchJob(
-                cache,
-                objectFetcher,
+    /**
+     * Creates fetch requests for a single object with multiple batches.
+     * Aligns byte ranges and aggregates metadata.
+     */
+    private Stream<ObjectFetchRequest> createFetchRequests(
+        final String objectKey,
+        final List<MergedBatchRequest> batches
+    ) {
+        // Get the most recent timestamp (for hot/cold decision)
+        final long timestamp = batches.stream()
+            .mapToLong(MergedBatchRequest::timestamp)
+            .max()
+            .orElse(0L);
+
+        // Sum up all byte sizes (for rate limiting)
+        final long totalBytes = batches.stream()
+            .mapToLong(MergedBatchRequest::byteSize)
+            .sum();
+
+        // Align byte ranges for efficient fetching
+        final Set<ByteRange> alignedRanges = keyAlignment.align(
+            batches.stream()
+                .map(MergedBatchRequest::byteRange)
+                .collect(Collectors.toList())
+        );
+
+        // Create a fetch request for each aligned range
+        return alignedRanges.stream()
+            .map(byteRange -> new ObjectFetchRequest(
                 objectKeyCreator.from(objectKey),
                 byteRange,
-                time,
-                metrics::fetchFileFinished,
-                metrics::cacheEntrySize
+                timestamp,
+                totalBytes
```

Comment:
If total bytes will be used for rate limiting there will be a sum calculated further out of these values, right? But here every request will get the totalBytes and will have the same value, should it be `byteRange.size()` instead? Or could you clarify what am I missing?

Reply:
I'm removing the ts/bytes from this PR as it's mixing concerns from a following PR here. I've refactored the changes on the first fixup commit to simplify this.